### ISSUE TITLE
bin/buildout (2.0.1) fails with Unicode named directories in dist-packages

### ISSUE THREAD

**Issue by jandd** [buildout/buildout#98]
I use zc.buildout  for a Django project on Debian (Wheezy). I just ran into trouble on a machine that has the system package python-pyramid installed. The package contains a file named `/usr/lib/python2.7/dist-packages/pyramid/tests/fixtures/static/héhé` (the accented `é`s are stored UTF-8 encoded on the file system). This leads to the following problem with running bin/buildout:

```
While:
  Installing.

An internal error occured due to a bug in either zc.buildout or in a
recipe being used:
Traceback (most recent call last):
  File "/home/jan/.buildout/eggs/zc.buildout-2.0.1-py2.7.egg/zc/buildout/buildout.py", line 1808, in main
    getattr(buildout, command)(args)
  File "/home/jan/.buildout/eggs/zc.buildout-2.0.1-py2.7.egg/zc/buildout/buildout.py", line 505, in install
    self._compute_part_signatures(install_parts)
  File "/home/jan/.buildout/eggs/zc.buildout-2.0.1-py2.7.egg/zc/buildout/buildout.py", line 749, in _compute_part_signatures
    sig = _dists_sig(pkg_resources.working_set.resolve([req]))
  File "/home/jan/.buildout/eggs/zc.buildout-2.0.1-py2.7.egg/zc/buildout/buildout.py", line 1515, in _dists_sig
    result.append(dist.project_name + '-' + _dir_hash(location))
  File "/home/jan/.buildout/eggs/zc.buildout-2.0.1-py2.7.egg/zc/buildout/buildout.py", line 1487, in _dir_hash
    hash.update(' '.join(dirnames).encode())
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 1: ordinal not in range(128)
```

Uninstalling python-pyramid helps (running python -S bootstrap.py / python -S bin/buildout does not). Uninstalling python-pyramid is not really an option because there are packaged system wide available applications that require it.

**Comment by idgserpro** [comment]
@reinout Shouldn't we try to close at least two issues relating unicode problems? We have this one, https://github.com/buildout/buildout/issues/148 and https://github.com/buildout/buildout/issues/152.

Fixing it would be really nice in a next release (https://github.com/buildout/buildout/issues/249). This is really, really confusing for beginners. Even an unicode named directory in ```~/.local``` breaks zc.buildout.

**Comment by reinout** [comment]
A fix might be to use a function like the `fs_to_text()` as in zest.releaser. We started supporting python 3 a few weeks ago and this was one of the helper methods that came of of that.

See https://github.com/zestsoftware/zest.releaser/blob/master/zest/releaser/utils.py#L44

**Comment by reinout** [comment]
Pull request #250 fixes this. Can someone take a look?

**Comment by idgserpro** [comment]
@reinout Thank you very much for the fix. This bug is *really* obscure to people new to Plone - if you try to install ```pip install --user mr.bob``` for example, an egg suggested for [Plone 5 Training](https://plone-training.readthedocs.org/en/latest/eggs1.html) and seen as harmless, it just breaks the buildout. [Here's the culprit in mr.bob](https://github.com/domenkozar/mr.bob/blob/master/mrbob/tests/templates/encodingc%CC%8C/mapc%CC%8Ca/c%CC%81a.bob).

```
buildout.py", line 1605, in _dir_hash
    hash.update(' '.join(dirnames).encode())

UnicodeDecodeError: 'ascii' codec can't decode byte 0xcc in position 31: ordinal not in range(128)

['ds_store', 'empty', 'empty2', 'encodingc\xcc\x8c', 'encoding\xc4\x8d', 'ignored', 'ignored_dirs', 'missing_namespace_key', 'multiconfig', 'questions1', 'questions2', 'questions3', 'questions4', 'questions5', 'renamed', 'renamedfile', 'renamedir', 'renamedtemplate', 'renamedtemplate2', 'render_hooks', 'renderer', 'skip_mrbobini', 'unbound']
```

I created an egg from your branch to use in a local environment and tested in the ```~/.local``` scenario and it works in an Ubuntu workstation 12.04, Python 2.7.9. @jandd could you test again with the ```python-pyramid``` scenario?

**Comment by reinout** [comment]
@idgserpro: thanks, that's a pretty good test that my fix works :-)

**Comment by idgserpro** [comment]
You can try to mess up with your installation and do the same test. :) It works with any buildout you run after installing mr.bob in ~/.local using ```pip install --user```. I believe it could be tested on a mac if you have it. Let's wait for @jandd for the other situation.

**Comment by reinout** [comment]
@idgserpro can you try it out the same way with #252? That's a much smaller, simpler fix than my original one.

**Comment by idgserpro** [comment]
We will try to take a look. Don't forget to, after merging, update the bootstrap branch (thanks for the docs about bootstrap.py lifecycle BTW). After merging I think https://github.com/buildout/buildout/issues/86 can be closed as well.

**Comment by idgserpro** [comment]
Here is the test made:

- Run buildout without mr.bob installed. It works well;
- Install mr.bob using ```pip install --user```. Run buildout. Error ```UnicodeDecodeError: 'ascii' codec can't decode byte 0xcc in position 31: ordinal not in range(128)``` occurs;
- Uninstall mr.bob using ```pip uninstall mr.bob```. Run buildout. It works well;
- Reinstall mr.bob using ```pip install --user```. Run buildout. Error ```UnicodeDecodeError: 'ascii' codec can't decode byte 0xcc in position 31: ordinal not in range(128)``` happens again;
- Create a new *internal* release from the ```reinout-simpler-250-fix``` branch, **2.4.0.dev0**, and pin my package to download ```zc.buildout = 2.4.0.dev0``` from our pypiserver. Run buildout. No error occurs.

When creating new release 2.4.0, I will test again.

**Comment by reinout** [comment]
Buildout 2.4.0 fixes it.
